crates/router-controller/src/controller.rs

```diff
@@ -127,8 +131,31 @@ impl Controller {
             0 => None,
             1 => Some(&candidates[0].url),
             len => {
-                let i: usize = rand::random();
-                Some(&candidates[i % len].url)
+                let key = (dataset.clone(), first_block);
+                candidates.sort_by_key(|info| self.workers_rate.get_rate(&info.url));
```

Comment:
Sort here by rate is not desired as we want a systematic cycling through workers in case of retries. Keeping the workers in the same order as we have them is fine.